### ISSUE TITLE
Why not keep the gumbel softmax trick during the retraining stage?

### ISSUE THREAD

**Issue by d12306** [D-X-Y/AutoDL-Projects#91]
Hello, @D-X-Y , thanks for your implementation, I noticed that NAS papers that employ the Gumbel-softmax trick in the searching stage do not keep the same sampling procedure in the evaluation stage. Why do you keep such an inconsistency between training and evaluation? 

Does adding the Gumbel sampling do any bad to the evaluation performance?

Thanks,

**Comment by D-X-Y** [comment]
Would you mind providing more details? What do you mean by "adding Gumbel sampling in the evaluation stage"? and which paper are you referring to?

**Comment by d12306** [comment]
I am referring paper such as GDAS, by saying "I noticed that NAS papers that employ the Gumbel-softmax trick in the searching stage do not keep the same sampling procedure in the evaluation stage.", I mean during the search, you sample from the Gumbel-softmax distribution and obtain the weight for different operations. And once you finish training, you use the latest weights of ops and get the architecture (finding the two most probable ops for each edge), however, Gumbel-softmax can be influenced by the random noise so the weights of ops are not deterministic given fixed logits. 

That being said, the architecture can be quite different if we repeatly sample from the gumbel-softmax distribution, what is the point that we always use one of them? (which can be regarded as a point estimate)

Thanks and please correct me if anything is wrong.

**Comment by D-X-Y** [comment]
@d12306 Sorry for the late response and thanks for the clarification.
The goal of differentiable is to learn the distribution of architectures (defined by a set of variables $alpha$). GDAS uses Gumbel-softmax to update this distribution. After searching, as this distribution has been optimized, we regard the op with the highest probability is the best.

**Comment by tehseenmayar** [comment]
Hi, i am sorry for asking such question because i am new to NAS and have been trying perform grid search over the search space of NAS Bench-201 but couldn't find a way so far. can you give me a head start about it?
I would appreciate any help

**Comment by D-X-Y** [comment]
@tehseenmayar Thanks for your interest.

We have extended our NAS-Bench-201 to [NATS-Bench](https://arxiv.org/abs/2009.00437), which has more architecture information and a more efficient and robust API. I would recommend you use NATS-Bench instead of NAS-Bench-201. If you want to start with
- random search, you can try this script: https://github.com/D-X-Y/AutoDL-Projects/blob/master/exps/NATS-algos/random_wo_share.py#L6
- REINFORCE, you can try this: https://github.com/D-X-Y/AutoDL-Projects/blob/master/exps/NATS-algos/reinforce.py#L6
- unfortunately, we did not provide codes for grid search, while it should be straightforward to modify from random search codes.

**Comment by tehseenmayar** [comment]
i tried to run this command but it seems like the api cannot find it. i am using NAS Bench-201.
from models import get_cell_based_tiny_net
I am trying to get the search space to make a dictionary and then to pass it to GridSearchCV or can you tell me how can i apply GridSearchCV over the search space to find the predicted best model?

**Comment by tehseenmayar** [comment]
Hi, Thanks for your great work. I have run the following command and got the results.
`python ./exps/NATS-algos/random_wo_share.py --dataset cifar100 --search_space sss`
can you tell me how to get the final discovered architecture?
I would appreciate your help.
Thanks

**Comment by D-X-Y** [comment]
![image](https://user-images.githubusercontent.com/9547057/106236602-36ae1100-6238-11eb-8876-bf449169a961.png)

Can you see something like this in the saved log file? "The best arch is xxx", where "xxx" is the final discovered architecture

**Comment by tehseenmayar** [comment]
<img width="936" alt="Screenshot 2021-01-29 at 12 55 25 PM" src="https://user-images.githubusercontent.com/29973339/106261012-21ce7f00-6232-11eb-91c2-840ae87f7a2e.png">
Thank you for your response. I am looking at these kinds of information. How to get the informations which you have mentioned above. I need to train the found architecture for accuracy validation.

**Comment by tehseenmayar** [comment]
sorry to bother you again, can you also tell me about how can i retrain the final discovered architecture to find the validation accuracy of that architecture?

thank you

**Comment by D-X-Y** [comment]
If you are using our NAS-Bench-201 or NATS-Bench (https://xuanyidong.com/assets/projects/NATS-Bench), you do not need to re-train the model. You can directly query the performance of your discovered information via our API.
